ui/impl/src/main/java/org/jboss/forge/addon/ui/impl/controller/CacheCommandControllerDecorator.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2017 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Eclipse Public License version 1.0, available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.jboss.forge.addon.ui.impl.controller;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.jboss.forge.addon.ui.controller.CommandController;
+
+/**
+ *
+ * @author <a href="mailto:ggastald@redhat.com">George Gastaldi</a>
+ */
+public class CacheCommandControllerDecorator extends AbstractCommandControllerDecorator
+{
+   private final Map<String, Boolean> state = new ConcurrentHashMap<>(4);
+
+   public CacheCommandControllerDecorator(CommandController controller)
```

Comment:
Woah. What's this for?! Looks interesting :)

Reply:
Looks like this caches the states for commands being enabled/disabled to prevent re-computing the necessary facets/other determinants during command discovery?

Reply:
Yeah, not only during command discovery, but multiple validate() calls to determine the state of the current command for example

Reply:
Brilliant!! Now you have to worry about cache invalidation criterion ! But this is awesome.

Reply:
In this PR I assumed that the cache needs to be invalidated when setValue() is called, which would mean that the internal values changes and recomputation may be needed. Need to double-check why the tests are failing.